### DISCUSSION TITLE
Can table be saved when save panel?

### DISCUSSION THREAD

**Discussion by DRBAUS** [RomanKubiak/ctrlr#453] [Q&A]
When writing modulator values to a table can this table be saved in the panel itself so the next time the panel is loaded I can call a specific table and the modulator values are reloaded?

In that way modulators can be 're-used' and I could have a panel with just 1 Layer instead of many.

I noticed with the method provided by dnaldoog in another post works but the table is again empty when the panel is loaded again.

I am aware of the save to file method.

**Comment by damiensellier** [comment]
you can convert your table to a string and save it with stateData() . stateData() is a function in Ctrlr that creates a valuetree and saves/reads datas when the host launches or quits. You need to assign 2 scripts within the global propeties. No external file needed here. it's going to be added temporarly at the bottom of the Ctrlr.settings file in "user/library/preferences/Ctrlr/" on mac. And it is also saved as datas in your session file (in the xxxsessionnamexxx.cpr for cubase for exemple ) within a VST/AU host.

NOTE: It works only if the Ctrlr App is quited with the panel loaded. If you close the panel then quits it won't save the datas and could not load when you open Ctrlr App afterwards. it works very well within vst hosts. Everytime you save a session within the host it goes through the SaveStateData script and when you load the session it reads the ReadStateData script.
Ctrlr App behaves like a host.

![stateData](https://user-images.githubusercontent.com/55164385/160943383-372d4cf0-9f7d-4151-a4e4-9c6f23332337.png)



SaveStateData method  (global "called when Ctrlr State is saved") :

```
SaveStateData = function(--[[ ValueTree --]]stateData)

	if TableToSave ~= nil then  -- if the table exists or is not empty
		sTableToSave = table.concat(TableToSave, ", ") -- concatenate the table as a string with ", " as separator
		stateData:setProperty("DataToSave", sTableToSave, nil) -- save the string as "stateData"
	end

end -- f
```


ReadStateData method (global "called when Ctrlr is loaded"):
```
ReadStateData = function(--[[ ValueTree --]] stateData)

	if stateData:hasProperty("DataToSave") == true then  -- if the valuetree contains the string called DataToSave then
		sTableToSave = tostring(stateData:getProperty("DataToSave"))		 -- convert the data to a lua string
	
		TableToSave = {}  -- create an empty table called  TableToSave
		separator = ", "
		for sTableToSaveValue in string.gmatch(sTableToSave, "([^"..separator.."]+)") do   -- for each value separated by ", "
                    	table.insert(TableToSave, sTableToSaveValue)  -- insert the separated values in a table -- you now have your table back
                end -- for
    	end -- if

end -- f
```

I did not try it, it's based on some of my functions but it should work. Please let me know.

**Comment by dnaldoog** [comment]
You can actually save tables themselves using `stateData:addChild`:

<pre>
SaveStateData = function(--[[ ValueTree --]]stateData)
 if not rawequal(next(patchDataTable), nil) then -- it's probably never nil
        local mem = ValueTree("memory")
        for i, v in ipairs(patchDataTable) do
            mem:setProperty(tostring(v), tostring(panel:getModulatorByName(v):getModulatorValue()), nil) 
        end
        stateData:addChild(mem, -1, nil)
    end -- table not empty
end
</pre>

and reload like this:

<pre>
ReadStateData= function(--[[ ValueTree --]] stateData)
    local mi = stateData:getChildWithName("memory")
    patchDataTable = {}
    if mi ~= nil then -- ~ if stateData:hasProperty("mi") then ...
        for i = 0, mi:getNumProperties() - 1 do
            local k = mi:getPropertyName(i) -- k is the name of the control
            local v = mi:getProperty(k)
            patchDataTable[k] = v  -- and/or panel:getModulatorByName(k):setModularValue(v)
        end
    end
end
</pre>

in the `Ctrlr.settings` file the table would look like this:

`<panelCustomData><memory portSpeed="24" progVolume="22" osc1WaveformSaw="21" osc1Freq="20"/></panelCustomData>`


[record to table write to json + savestate_6_0.zip](https://github.com/RomanKubiak/ctrlr/files/8384738/record.to.table.write.to.json.%2B.savestate_6_0.zip)

**Reply by damiensellier** [reply]
that's excellent, really!
You are right; using json is so much better than lua table saved as a coma separated string.
With a "table to string" it requires an extra function to determine the number of elements within. 
With datas as Json you can use **getNumProperties** which returns the number of elements easily.
I didn't know that. As always John, thanks for the tip!

**Reply by dnaldoog** [reply]
@damiensellier  _Ctrlr_ is deep! 😸

**Comment by dnaldoog** [comment]
@DRBAUS 

I think this is what you are after.

There are 8 possible _wave sample_ values for each modulator, which are saved to state and can be saved as json.

----

Updated 04/04/2022

[record to table write to json + savestate_10_2.zip](https://github.com/RomanKubiak/ctrlr/files/8406137/record.to.table.write.to.json.%2B.savestate_10_2.zip)

![v10 1 0](https://user-images.githubusercontent.com/76831683/161431357-6a45b137-7139-47b6-98c7-4b2927bd6a64.png)

**Comment by dnaldoog** [comment]
This new version allows for spaces or dashes in the modulator name, but is much more complex coding wise:

----
[record to table write to json + savestate_11_3.zip](https://github.com/RomanKubiak/ctrlr/files/8447688/record.to.table.write.to.json.%2B.savestate_11_3.zip)
